### PR TITLE
supports training with airstore datasets

### PR DESCRIPTION
**Pull request by wpc**
I created a new type of dataset "airstore" on top of new AIRStore ioPath tabular interface. To make the check point resuming work, I need replicate the logic that set_epoch and set_start_iter on sampler, do the same thing for dataset objs.
I have tested it in faircluster
* doc for setup the integration:  https://docs.google.com/document/d/1RNSjJcFTGl4Or-9SRLYbB-ffuOMidZy62SwBA9L0Qm0/edit?usp=sharing
* example output (imagenet with config quick_simclr_2node): /checkpoint/wpc/vissl/2021-04-14-17-03-36/checkpoints/log.txt

### CODE REVIEW

File: vissl/data/airstore_dataset.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            rank=self.global_rank * num_workers + worker_id, # each worker take it's split by it's parent process rank and worker id
+            limit=getattr(split_cfg, "DATA_LIMIT", -1),
+            offset=getattr(split_cfg, "DATA_OFFSET", 0),
+            num_of_threads=getattr(split_cfg, "AIRSTORE_NUM_OF_THREADS", 2),
+            prefetch=getattr(split_cfg, "AIRSTORE_PREFETCH", 1),
+            max_holding_bundles=getattr(split_cfg, "AIRSTORE_MAX_HOLDING_BUNDLES", 5),
+            bundle_download_timeout_ms=getattr(split_cfg, "AIRSTORE_BUNDLE_DOWNLOAD_TIMEOUT_MS", 30000),
+            max_retries=getattr(split_cfg, "AIRSTORE_MAX_RETRIES", 5),
+            dataset_catalog_path=getattr(split_cfg, "AIRSTORE_DS_CATALOG_PATH", None), # temporary need during airstore development
+            env="OSS",
```
Comment (prigoyal) [inline vissl/data/airstore_dataset.py:81]: can we add a comment to explain the "OSS" ? :)

File: vissl/data/airstore_dataset.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+    def __getitem__(self, index):
+        if self._iterator is None:
+            self._iterator = self._open_iterator()
+
+        if not self.queue_init and self.enable_queue_dataset:
+            self._init_queues()
+
+        try:
+            # TODO (wpc, prigoyal): we should check images are good when we are
```
Comment (prigoyal) [inline vissl/data/airstore_dataset.py:100]: it would be great to verify this now. Essentially, when we get images from fairstore, we observed the images could be truncated . Correctness of images, if possible, could be tested during the ingestion time in airstore. Thoughts? :)

File: vissl/data/ssl_dataset.py [hunk 3]
```diff
@@ -118,6 +118,16 @@ def _get_data_limit_sampling(cfg: AttrDict, split: str) -> AttrDict:
         )
         return cfg["DATA"][split].get("DATA_LIMIT_SAMPLING", default_sampling)
 
+    def set_epoch(self, epoch):
```
Comment (prigoyal) [inline vissl/data/ssl_dataset.py:121]: hmm this becomes a little confusing since it's special case for the Airstore dataset only. We should add:
1) a clear comment to this function explaining why we set_epoch for airstore and why this isn't require for non-airstore datasets 
2) assert to check the data is airstore

File: vissl/data/ssl_dataset.py [hunk 4]
```diff
@@ -118,6 +118,16 @@ def _get_data_limit_sampling(cfg: AttrDict, split: str) -> AttrDict:
...
+        for data_obj in self.data_objs:
+            if hasattr(data_obj, "set_epoch"):
+                data_obj.set_epoch(epoch)
+
+    def set_start_iter(self, start_iter):
```
Comment (prigoyal) [inline vissl/data/ssl_dataset.py:126]: same comment as above

File: vissl/data/ssl_dataset.py [hunk 5]
```diff
@@ -118,6 +118,16 @@ def _get_data_limit_sampling(cfg: AttrDict, split: str) -> AttrDict:
         )
         return cfg["DATA"][split].get("DATA_LIMIT_SAMPLING", default_sampling)
 
+    def set_epoch(self, epoch):
+        for data_obj in self.data_objs:
+            if hasattr(data_obj, "set_epoch"):
+                data_obj.set_epoch(epoch)
```
Comment (prigoyal) [inline vissl/data/ssl_dataset.py:124]: I think this will fail here if the data_obj is not airstore object? since we don't have the set_epoch implemented for other datasets.

File: vissl/hooks/state_update_hooks.py [hunk 6]
```diff
@@ -95,6 +95,11 @@ def on_phase_start(self, task: "tasks.ClassyTask") -> None:
             if hasattr(task.dataloaders[phase_type].sampler, "set_epoch"):
                 # task.phase_idx is current running phase id
                 task.dataloaders[phase_type].sampler.set_epoch(task.phase_idx)
+        if hasattr(task.dataloaders[phase_type], "dataset"):
+            if hasattr(task.dataloaders[phase_type].dataset, "set_epoch"):
```
Comment (prigoyal) [inline vissl/hooks/state_update_hooks.py:99]: how does this work with non-airstore datasets? will the conditional be still satisfied since dataset is "GenericSSLDataset" which has the `set_epoch` ?

File: vissl/trainer/train_task.py [hunk 7]
```diff
@@ -506,6 +506,27 @@ def recreate_data_iterator(self, phase_type, epoch, compute_start_iter):
                     start_iter = 0
                 sampler.set_start_iter(start_iter)
             print_sampler_config(sampler)
+
+        if hasattr(self.dataloaders[phase_type], "dataset"):
```
Comment (prigoyal) [inline vissl/trainer/train_task.py:510]: same comment as earlier

**Review by prigoyal — COMMENTED** [review]
thank you so much @wpc , I am very excited about this :) 

- I left some inline comments that would be good to address. 
- It would be good to run a full scale test on FAIR cluster to verify the loss curves with disk and airstore imagenet
```
./dev/launch_slurm.sh \
    config=pretrain/swav/swav_8node_resnet \
    config.OPTIMIZER.num_epochs=100 \
    config.OPTIMIZER.param_schedulers.lr.lengths=[0.1, 0.9] \
    config.DATA.TRAIN.COPY_TO_LOCAL_DISK=False \ 
    config.SLURM.NAME=swav_100ep_rn50_in1k \
    config.SLURM.PARTITION=dev
```

- If you want, you can add `imagenet_airstore` to the `dataset_catalog.json`

File: vissl/data/__init__.py [hunk 8]
```diff
@@ -33,6 +34,7 @@
 ]
 
 DATASET_SOURCE_MAP = {
+    "airstore": AirstoreDataset,
```
[the remark this answers is not in the document]
Comment (wpc) [reply vissl/data/__init__.py:37]: yes, I have a next diff working on it

File: vissl/data/airstore_dataset.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+            world_size=self.global_world_size * num_workers, # split the dataset for each worker
+            rank=self.global_rank * num_workers + worker_id, # each worker take it's split by it's parent process rank and worker id
+            limit=getattr(split_cfg, "DATA_LIMIT", -1),
+            offset=getattr(split_cfg, "DATA_OFFSET", 0),
+            num_of_threads=getattr(split_cfg, "AIRSTORE_NUM_OF_THREADS", 2),
+            prefetch=getattr(split_cfg, "AIRSTORE_PREFETCH", 1),
+            max_holding_bundles=getattr(split_cfg, "AIRSTORE_MAX_HOLDING_BUNDLES", 5),
+            bundle_download_timeout_ms=getattr(split_cfg, "AIRSTORE_BUNDLE_DOWNLOAD_TIMEOUT_MS", 30000),
+            max_retries=getattr(split_cfg, "AIRSTORE_MAX_RETRIES", 5),
+            dataset_catalog_path=getattr(split_cfg, "AIRSTORE_DS_CATALOG_PATH", None), # temporary need during airstore development
```
[the remark this answers is not in the document]
Comment (wpc) [reply vissl/data/airstore_dataset.py:80]: AIRSTORE_DS_CATALOG_PATH is for finding airstore dataset in faircluster in a CSV file. Coincidentally it is also called dataset_catalog :-).

File: vissl/data/airstore_dataset.py (hunk 2 above)
Comment (wpc) [reply vissl/data/airstore_dataset.py:100]: I totally agree this should be handled at the ingestion time. let me talk with the team

File: vissl/data/ssl_dataset.py (hunk 5 above)
Comment (wpc) [reply vissl/data/ssl_dataset.py:124]: It has "if hasattr(data_obj, "set_epoch"):" check. I am thinking make the set_epoch an optional interface on dataset by duck typing. It make sense for all future iterator based dataset implementations, where they can not rely on sampler to do the checkpoint recovering. If we don't plan to have another one, we can hardcode it.

File: vissl/hooks/state_update_hooks.py (hunk 6 above)
Comment (wpc) [reply vissl/hooks/state_update_hooks.py:99]: Yes, GenericSSLDataset will go through each sub dataset objs, and if it implemented "set_epoch" it will dispatch down

File: vissl/data/airstore_dataset.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+            queue_size=cfg["DATA"][split]["BATCHSIZE_PER_REPLICA"]
+        )
+        self.pathmgr = create_path_manager()
+        self.cfg = cfg
+        self.batch_size = cfg["DATA"][split]["BATCHSIZE_PER_REPLICA"]
+        self.airstore_uri = path
+        self.split = split
+        self.epoch = 0
+        self.start_iter = 0
+        self.enable_queue_dataset = cfg["DATA"][self.split]["ENABLE_QUEUE_DATASET"]
```
[the remark this answers is not in the document]
Comment (wpc) [reply vissl/data/airstore_dataset.py:33]: yes, tested in fb side (f266302140)

File: vissl/data/ssl_dataset.py (hunk 3 above)
Comment (wpc) [reply vissl/data/ssl_dataset.py]: recent change directly checks dataset is an instance of AirstoreDataSet, so set_epoch and set_start_iter don't need be added to GenericSSLDataset any more

File: vissl/data/ssl_dataset.py (hunk 5 above)
Comment (wpc) [reply vissl/data/ssl_dataset.py]: I change the code following the suggestions, now it directly checking the dataset type

File: vissl/hooks/state_update_hooks.py (hunk 6 above)
Comment (wpc) [reply vissl/hooks/state_update_hooks.py]: I change the code directly check dataset types

**Comment by wpc** [comment]
Did some test as Priya suggested:

Baseline swav_8node_resnet_100ep_disk_imagenet:
```
 ./dev/launch_slurm.sh \
    config=pretrain/swav/swav_8node_resnet \
    config.OPTIMIZER.num_epochs=100 \
    config.OPTIMIZER.param_schedulers.lr.lengths=[0.1,\ 0.9] \
    config.DATA.TRAIN.COPY_TO_LOCAL_DISK=False \
    config.SLURM.NAME=swav_100ep_rn50_in1k \
    config.SLURM.PARTITION=learnfair
```
Output dir: /checkpoint/wpc/vissl/2021-04-18-06-32-50/checkpoints/

Compare swav_8node_resnet_100ep_airstore_imagenet
```
./dev/launch_slurm.sh \
    config=pretrain/swav/swav_8node_resnet \
    config.OPTIMIZER.num_epochs=100 \
    config.OPTIMIZER.param_schedulers.lr.lengths=[0.1,\ 0.9] \
    config.DATA.TRAIN.COPY_TO_LOCAL_DISK=False \
    config.SLURM.NAME=swav_100ep_rn50_in1k \
    config.SLURM.PARTITION=learnfair \
    config.DATA.TRAIN.DATASET_NAMES=[airstore_imagenet] \
    config.DATA.TRAIN.DATA_SOURCES=[airstore] \
 +config.DATA.TRAIN.AIRSTORE_DS_CATALOG_PATH="$HOME/projects/AIRStore/airstore/data/airstore_test_dataset_catalog.csv"
```
Output dir: /checkpoint/wpc/vissl/2021-04-18-05-53-54/checkpoints/

Loss curve: 
download-5

btime:
download-6

**Comment by wpc** [comment]
Also post the iteration based checkpoint resuming  verification:

Testing setup:

2 node, imagenet (1.2M images), 2 epoch, total 10008 iterations

Command start training:
```
(vissl_test) wpc@devfair0174:~/code/vissl$ ./dev/launch_slurm.sh     config=test/integration_test/quick_simclr_2node     config.SLURM.NAME=quick_simclr_2ep_inairstore     config.SLURM.PARTITION=dev config.DATA.TRAIN.DATASET_NAMES=[airstore_imagenet]     config.DATA.TRAIN.DATA_SOURCES=[airstore]     +config.DATA.TRAIN.AIRSTORE_DS_CATALOG_PATH="$HOME/projects/AIRStore/airstore/data/airstore_test_dataset_catalog.csv" config.CHECKPOINT.CHECKPOINT_ITER_FREQUENCY=310 config.DATA.TRAIN.USE_STATEFUL_DISTRIBUTED_SAMPLER=true
```

Command for resume testing 
```
(vissl_test) wpc@devfair0174:~/code/vissl$ CHECKPOINT_DIR={checkpoint dir from previous incomplete run} ./dev/launch_slurm.sh     config=test/integration_test/quick_simclr_2node     config.SLURM.NAME=quick_simclr_2ep_inairstore     config.SLURM.PARTITION=dev config.DATA.TRAIN.DATASET_NAMES=[airstore_imagenet]     config.DATA.TRAIN.DATA_LIMIT=-1 config.DATA.TRAIN.DATA_SOURCES=[airstore]     +config.DATA.TRAIN.AIRSTORE_DS_CATALOG_PATH="$HOME/projects/AIRStore/airstore/data/airstore_test_dataset_catalog.csv" config.CHECKPOINT.CHECKPOINT_ITER_FREQUENCY=310 config.DATA.TRAIN.USE_STATEFUL_DISTRIBUTED_SAMPLER=true
```
Testing scenario:

Start a train:
EXP_ROOT_DIR: /checkpoint/wpc/vissl/2021-04-18-03-37-45
CHECKPOINT_DIR: /checkpoint/wpc/vissl/2021-04-18-03-37-45/checkpoints/

After a while kill the job using scancel then resume:
EXP_ROOT_DIR: /checkpoint/wpc/vissl/2021-04-18-03-58-14
CHECKPOINT_DIR: /checkpoint/wpc/vissl/2021-04-18-03-37-45/checkpoints

After a while kill again then resume:
EXP_ROOT_DIR: /checkpoint/wpc/vissl/2021-04-18-04-22-13
CHECKPOINT_DIR: /checkpoint/wpc/vissl/2021-04-18-03-37-45/checkpoints

Result:

1) Via airstore verbose log we can see on resuming the download bundle index starting from where the last checkpoint left instead of 0.
```
I0418 03:59:21.974376 1557142 StorageDaemonClient.cpp:222] send download request for bundle index 77 ongoing manifold download: 2
I0418 03:59:21.974380 1557139 StorageDaemonClient.cpp:222] send download request for bundle index 78 ongoing manifold download: 2
```

Compare with baseline, same training without checkpoint resuming in /checkpoint/wpc/vissl/2021-04-18-03-37-23/checkpoints/

Lost curv matches
download-4

File: vissl/hooks/state_update_hooks.py [hunk 11]
```diff
@@ -95,6 +96,17 @@ def on_phase_start(self, task: "tasks.ClassyTask") -> None:
             if hasattr(task.dataloaders[phase_type].sampler, "set_epoch"):
                 # task.phase_idx is current running phase id
                 task.dataloaders[phase_type].sampler.set_epoch(task.phase_idx)
+
+        # call set_epoch and for AirstoreDataset since it handles shuffle
```
Comment (prigoyal) [inline vissl/hooks/state_update_hooks.py:100]: nice! :) thank you :)

File: vissl/hooks/state_update_hooks.py [hunk 12]
```diff
@@ -95,6 +96,17 @@ def on_phase_start(self, task: "tasks.ClassyTask") -> None:
...
+        # behavior internally
+        if hasattr(task.dataloaders[phase_type], "dataset"):
+            dataset = task.dataloaders[phase_type].dataset
+            if isinstance(dataset, GenericSSLDataset):
+                for data_obj in dataset.data_objs:
+                    if isinstance(data_obj, AirstoreDataset):
```
Comment (prigoyal) [inline vissl/hooks/state_update_hooks.py:106]: nice! thank you.

File: vissl/hooks/state_update_hooks.py [hunk 13]
```diff
@@ -95,6 +96,17 @@ def on_phase_start(self, task: "tasks.ClassyTask") -> None:
             if hasattr(task.dataloaders[phase_type].sampler, "set_epoch"):
                 # task.phase_idx is current running phase id
                 task.dataloaders[phase_type].sampler.set_epoch(task.phase_idx)
+
+        # call set_epoch and for AirstoreDataset since it handles shuffle
+        # behavior internally
+        if hasattr(task.dataloaders[phase_type], "dataset"):
```
Comment (prigoyal) [inline vissl/hooks/state_update_hooks.py:102]: question: do we want to test the phase_type=train or test? aka, there's no strict requirement for the `phase_type=train` only?

**Review by prigoyal — APPROVED** [review]
this looks great and good to go to me :) thank you so so much. The test plan looks amazing and great to see the overlapping curves :)

File: vissl/hooks/state_update_hooks.py (hunk 13 above)
Comment (wpc) [reply vissl/hooks/state_update_hooks.py:102]: yeah I think check phase_type=train make sense. I already clicked the land button. Will create a new PR for it
